packages/apollo-link/src/types.ts

```diff
@@ -4,16 +4,13 @@ import { ExecutionResult, DocumentNode } from 'graphql';
 export interface GraphQLRequest {
   query?: string | DocumentNode;
   variables?: Record<string, any>;
-  context?: Record<string, any>;
-}
-
-export interface Operation {
-  query: DocumentNode;
```

Comment:
I don't really understand why we don't want the AST here - how could someone write a link that inspects the contents of the query in the new world?

Reply:
I think its reasonable to create a version of apollo-client / apollo-link that operations solely on a backend cache to manage updates instead of having the AST on the client.

The client would send an id of the operation + variables and would get back the resulting data and a set of queryIds that need to be invalidated (and possible the new data). The store would be a pretty dumb / low level Map.

Essentially I think that `query` should be a required field, but that it should be `string | DocumentNode`

Reply:
> I think its reasonable to create a version of apollo-client / apollo-link that operations solely on a backend cache to manage updates instead of having the AST on the client.

I feel that this is too far in the future to consider in this version of Link; if we make the implementation too generic that makes it less useful IMO.

> how could someone write a link that inspects the contents of the query in the new world?

Still wondering about this.

Reply:
I think I agree with James on the move to using a `string` for the `query` field. I was very worried that the move would make splitting on query properties very difficult and a `string` would be too generic.

The positives of having a string is that it makes persistent queries possible without additional effort. We don't have to include the printer/parser for every link stack.

The downside is additional trouble with DocumentNode specific behavior, such as splitting on subscriptions vs mutations/queries. My impression is that this will be a common feature that advanced solutions will need, so adding overhead is okay. The solution is to create a parsing link that adds the DocumentNode and possibly information about the operation type to the context. It also provides people the flexibility to define their own metadata around what a string query does.

Another issue that moving to strings causes is lack of error handling around an invalid query during runtime. This should ideally be done by codegen or some linting software if the queries are static, which leads into the next issue. Using a string means difficulty adding fields to a query, such as `__typename`. Not quite sure how to make this easier yet.

Another option is to make the query an object. This helps with changing fields and introspection. The translation to a network request would be more difficult.